File: rust/DataStruct/tree/avl_tree.rs
use std::cell::RefCell;
use std::cell::Cell;
use std::rc::Rc;
use std::cmp::Ordering;

#[derive(Debug)]
struct TreeNode {
    val: i32,
    height: i32,
    left: Option<Rc<RefCell<TreeNode>>>,
    right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    fn new(val: i32) -> Rc<RefCell<TreeNode>> {
        Rc::new(RefCell::new(Self {
            val,
            height: 0,
            left: None,
            right: None,
        }))
    }
}

type OptionTreeNodeRc = Option<Rc<RefCell<TreeNode>>>;

struct AVLTree {
    root: OptionTreeNodeRc,
}

impl AVLTree {
    fn new() -> Self {
        Self { root: None }
    }

    fn height(node: OptionTreeNodeRc) -> i32 {
        match node {
            Some(node) => node.borrow().height,
            None => -1,
        }
    }

    fn update_height(node: OptionTreeNodeRc) {
        if let Some(node) = node {
            let left = node.borrow().left.clone();
            let right = node.borrow().right.clone();
            node.borrow_mut().height = std::cmp::max(Self::height(left), Self::height(right)) + 1;
        }
    }

    fn balance_factor(node: OptionTreeNodeRc) -> i32 {
        match node {
            None => 0,
            Some(node) => {
                Self::height(node.borrow().left.clone()) - Self::height(node.borrow().right.clone())
            }
        }
    }

    fn right_rotate(node: OptionTreeNodeRc) -> OptionTreeNodeRc {
        match node {
            Some(node) => {
                let child = node.borrow().left.clone().unwrap();
                let grand_child = child.borrow().right.clone();
                child.borrow_mut().right = Some(node.clone());
                node.borrow_mut().left = grand_child;
                Self::update_height(Some(node));
                Self::update_height(Some(child.clone()));
                Some(child)
            }
            None => None,
        }
    }

    fn left_rotate(node: OptionTreeNodeRc) -> OptionTreeNodeRc {
        match node {
            Some(node) => {
                let child = node.borrow().right.clone().unwrap();
                let grand_child = child.borrow().left.clone();
                child.borrow_mut().left = Some(node.clone());
                node.borrow_mut().right = grand_child;
                Self::update_height(Some(node));
                Self::update_height(Some(child.clone()));
                Some(child)
            }
            None => None,
        }
    }

    fn rotate(node: OptionTreeNodeRc) -> OptionTreeNodeRc {
        let balance_factor = Self::balance_factor(node.clone());
        if balance_factor > 1 {
            let node = node.unwrap();
            if Self::balance_factor(node.borrow().left.clone()) >= 0 {
                Self::right_rotate(Some(node))
            } else {
                let left = node.borrow().left.clone();
                node.borrow_mut().left = Self::left_rotate(left);
                Self::right_rotate(Some(node))
            }
        } else if balance_factor < -1 {
            let node = node.unwrap();
            if Self::balance_factor(node.borrow().right.clone()) <= 0 {
                Self::left_rotate(Some(node))
            } else {
                let right = node.borrow().right.clone();
                node.borrow_mut().right = Self::right_rotate(right);
                Self::left_rotate(Some(node))
            }
        } else {
            node
        }
    }

    fn insert(&mut self, val: i32) {
        self.root = Self::insert_helper(self.root.clone(), val);
    }

    fn insert_helper(node: OptionTreeNodeRc, val: i32) -> OptionTreeNodeRc {
        match node {
            Some(mut node) => {
                match {
                    let node_val = node.borrow().val;
                    node_val
                }
                .cmp(&val)
                {
                    Ordering::Greater => {
                        let left = node.borrow().left.clone();
                        node.borrow_mut().left = Self::insert_helper(left, val);
                    }
                    Ordering::Less => {
                        let right = node.borrow().right.clone();
                        node.borrow_mut().right = Self::insert_helper(right, val);
                    }
                    Ordering::Equal => {
                        return Some(node);
                    }
                }
                Self::update_height(Some(node.clone()));
                node = Self::rotate(Some(node)).unwrap();
                Some(node)
            }
            None => Some(TreeNode::new(val)),
        }
    }

    fn remove(&self, val: i32) {
        Self::remove_helper(self.root.clone(), val);
    }

    fn remove_helper(node: OptionTreeNodeRc, val: i32) -> OptionTreeNodeRc {
        match node {
            Some(mut node) => {
                if val < node.borrow().val {
                    let left = node.borrow().left.clone();
                    node.borrow_mut().left = Self::remove_helper(left, val);
                } else if val > node.borrow().val {
                    let right = node.borrow().right.clone();
                    node.borrow_mut().right = Self::remove_helper(right, val);
                } else if node.borrow().left.is_none() || node.borrow().right.is_none() {
                    let child = if node.borrow().left.is_some() {
                        node.borrow().left.clone()
                    } else {
                        node.borrow().right.clone()
                    };
                    match child {
                        None => {
                            return None;
                        }
                        Some(child) => node = child,
                    }
                } else {
                    let mut temp = node.borrow().right.clone().unwrap();
                    loop {
                        let temp_left = temp.borrow().left.clone();
                        if temp_left.is_none() {
                            break;
                        }
                        temp = temp_left.unwrap();
                    }
                    let right = node.borrow().right.clone();
                    node.borrow_mut().right = Self::remove_helper(right, temp.borrow().val);
                    node.borrow_mut().val = temp.borrow().val;
                }
                Self::update_height(Some(node.clone()));

                node = Self::rotate(Some(node)).unwrap();
                Some(node)
            }
            None => None,
        }
    }

    fn search(&self, val: i32) -> OptionTreeNodeRc {
        let mut cur = self.root.clone();
        while let Some(node) = cur.clone() {
            match val.cmp(&node.borrow().val) {
                Ordering::Greater => {
                    cur = node.borrow().right.clone();
                }
                Ordering::Less => {
                    cur = node.borrow().left.clone();
                }
                Ordering::Equal => {
                    break;
                }
            }
        }
        cur
    }
}

struct Trunk<'a, 'b> {
    prev: Option<&'a Trunk<'a, 'b>>,
    str: Cell<&'b str>,
}

/* 打印二叉树 */
pub fn print_tree(root: &Rc<RefCell<TreeNode>>) {
    _print_tree(Some(root), None, false);
}

/* 打印二叉树 */
fn _print_tree(root: Option<&Rc<RefCell<TreeNode>>>, prev: Option<&Trunk>, is_right: bool) {
    if let Some(node) = root {
        let mut prev_str = "    ";
        let trunk = Trunk { prev, str: Cell::new(prev_str) };
        _print_tree(node.borrow().right.as_ref(), Some(&trunk), true);

        if  prev.is_none() {
            trunk.str.set("———");
        } else if is_right {
            trunk.str.set("/———");
            prev_str = "   |";
        } else {
            trunk.str.set("\\———");
            prev.as_ref().unwrap().str.set(prev_str);
        }

        show_trunks(Some(&trunk));
        println!(" {}", node.borrow().val);
        if let Some(prev) = prev {
            prev.str.set(prev_str);
        }
        trunk.str.set("   |");

        _print_tree(node.borrow().left.as_ref(), Some(&trunk), false);
    }
}

fn show_trunks(trunk: Option<&Trunk>) {
    if let Some(trunk) = trunk {
        show_trunks(trunk.prev);
        print!("{}", trunk.str.get());
    }
}




fn main() {
    fn test_insert(tree: &mut AVLTree, val: i32) {
        tree.insert(val);
        println!("\n插入节点 {} 后，AVL 树为", val);
        print_tree(&tree.root.clone().unwrap());
    }

    fn test_remove(tree: &mut AVLTree, val: i32) {
        tree.remove(val);
        println!("\n删除节点 {} 后，AVL 树为", val);
        print_tree(&tree.root.clone().unwrap());
    }

    /* 初始化空 AVL 树 */
    let mut avl_tree = AVLTree::new();

    /* 插入节点 */
    // 请关注插入节点后，AVL 树是如何保持平衡的
    test_insert(&mut avl_tree, 1);
    test_insert(&mut avl_tree, 2);
    test_insert(&mut avl_tree, 3);
    test_insert(&mut avl_tree, 4);
    test_insert(&mut avl_tree, 5);
    test_insert(&mut avl_tree, 8);
    test_insert(&mut avl_tree, 7);
    test_insert(&mut avl_tree, 9);
    test_insert(&mut avl_tree, 10);
    test_insert(&mut avl_tree, 6);

    /* 插入重复节点 */
    test_insert(&mut avl_tree, 7);

    /* 删除节点 */
    // 请关注删除节点后，AVL 树是如何保持平衡的
    test_remove(&mut avl_tree, 8); // 删除度为 0 的节点
    test_remove(&mut avl_tree, 5); // 删除度为 1 的节点
    test_remove(&mut avl_tree, 4); // 删除度为 2 的节点

    /* 查询节点 */
    let node = avl_tree.search(7);
    if let Some(node) = node {
        println!(
            "\n查找到的节点对象为 {:?}，节点值 = {}",
            &*node.borrow(),
            node.borrow().val
        );
    }
}

File: Cpp/DataStructure/AVlTree.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

struct TreeNode {
  int val{};
  int height = 0;
  TreeNode *parent{};
  TreeNode *left{};
  TreeNode *right{};
  TreeNode() = default;
  explicit TreeNode(int x, TreeNode *parent = nullptr)
      : val(x), parent(parent) {}
};

void freeMemoryTree(TreeNode *root) {
  if (root == nullptr) {
    return;
  }
  freeMemoryTree(root->left);
  freeMemoryTree(root->right);
  delete root;
}

class AVLTree {
private:
  void updateHeight(TreeNode *node) {
    node->height = std::max(height(node->left), height(node->right)) + 1;
  }
  /* 右旋操作 */
  TreeNode *rightRotate(TreeNode *node) {
    TreeNode *child = node->left;
    TreeNode *grandChild = child->right;
    // 以 child 为原点，将 node 向右旋转
    child->right = node;
    node->left = grandChild;
    // 更新节点高度
    updateHeight(node);
    updateHeight(child);
    // 返回旋转后子树的根节点
    return child;
  }

  /* 左旋操作 */
  TreeNode *leftRotate(TreeNode *node) {
    TreeNode *child = node->right;
    TreeNode *grandChild = child->left;
    // 以 child 为原点，将 node 向左旋转
    child->left = node;
    node->right = grandChild;
    // 更新节点高度
    updateHeight(node);
    updateHeight(child);
    // 返回旋转后子树的根节点
    return child;
  }

  /* 执行旋转操作，使该子树重新恢复平衡 */
  TreeNode *rotate(TreeNode *node) {
    // 获取节点 node 的平衡因子
    int _balanceFactor = balanceFactor(node);
    // 左偏树
    if (_balanceFactor > 1) {
      if (balanceFactor(node->left) >= 0) {
        // 右旋
        return rightRotate(node);
      } else {
        // 先左旋后右旋
        node->left = leftRotate(node->left);
        return rightRotate(node);
      }
    }
    // 右偏树
    if (_balanceFactor < -1) {
      if (balanceFactor(node->right) <= 0) {
        // 左旋
        return leftRotate(node);
      } else {
        // 先右旋后左旋
        node->right = rightRotate(node->right);
        return leftRotate(node);
      }
    }
    // 平衡树，无须旋转，直接返回
    return node;
  }

  TreeNode *insertHelper(TreeNode *node, int val) {
    if (node == nullptr)
      return new TreeNode(val);
    /* 1. 查找插入位置并插入节点 */
    if (val < node->val)
      node->left = insertHelper(node->left, val);
    else if (val > node->val)
      node->right = insertHelper(node->right, val);
    else
      return node;      // 重复节点不插入，直接返回
    updateHeight(node); // 更新节点高度
    /* 2. 执行旋转操作，使该子树重新恢复平衡 */
    node = rotate(node);
    // 返回子树的根节点
    return node;
  }

  TreeNode *removeHelper(TreeNode *node, int val) {
    if (node == nullptr) {
      return nullptr;
    }
    if (val < node->val) {
      node->left = removeHelper(node->left, val);
    } else if (val > node->val) {
      node->right = removeHelper(node->right, val);
    } else {
      if (node->left == nullptr || node->right == nullptr) {
        TreeNode *child = node->left != nullptr ? node->left : node->right;
        if (child == nullptr) {
          delete node;
          return nullptr;
        } else {
          delete node;
          node = child;
        }
      } else {
        TreeNode *temp = node->right;
        while (temp->left != nullptr) {
          temp = temp->left;
        }
        int tempVal = temp->val;
        node->right = removeHelper(node->right, tempVal);
        node->val = tempVal;
      }
    }
    updateHeight(node);
    node = rotate(node);
    return node;
  }

public:
  TreeNode *root;

  int height(TreeNode *node) { return node == nullptr ? -1 : node->height; }

  int balanceFactor(TreeNode *node) {
    if (node == nullptr) {
      return 0;
    }
    return height(node->left) - height(node->right);
  }

  void insert(int val) { root = insertHelper(root, val); }

  void remove(int val) { root = removeHelper(root, val); }

  TreeNode *search(int val) {
    TreeNode *cur = root;
    while (cur != nullptr) {
      if (cur->val < val) {
        cur = cur->right;
      } else if (cur->val > val) {
        cur = cur->left;
      } else {
        break;
      }
    }
    return cur;
  }

  AVLTree() : root(nullptr) {}

  ~AVLTree() { freeMemoryTree(root); }
};

struct Trunk {
  Trunk *prev;
  string str;
  Trunk(Trunk *prev, string str) {
    this->prev = prev;
    this->str = str;
  }
};

void showTrunks(Trunk *p) {
  if (p == nullptr) {
    return;
  }

  showTrunks(p->prev);
  cout << p->str;
}

/**
 * 打印二叉树
 * This tree printer is borrowed from TECHIE DELIGHT
 * https://www.techiedelight.com/c-program-print-binary-tree/
 */
void printTree(TreeNode *root, Trunk *prev, bool isRight) {
  if (root == nullptr) {
    return;
  }

  string prev_str = "    ";
  Trunk trunk(prev, prev_str);

  printTree(root->right, &trunk, true);

  if (!prev) {
    trunk.str = "———";
  } else if (isRight) {
    trunk.str = "/———";
    prev_str = "   |";
  } else {
    trunk.str = "\\———";
    prev->str = prev_str;
  }

  showTrunks(&trunk);
  cout << " " << root->val << endl;

  if (prev) {
    prev->str = prev_str;
  }
  trunk.str = "   |";

  printTree(root->left, &trunk, false);
}

/* 打印二叉树 */
void printTree(TreeNode *root) { printTree(root, nullptr, false); }

void testInsert(AVLTree &tree, int val) {
  tree.insert(val);
  std::cout << "\n插入节点 " << val << " 后，AVL 树为" << std::endl;
  printTree(tree.root);
}

void testRemove(AVLTree &tree, int val) {
  tree.remove(val);
  std::cout << "\n删除节点 " << val << " 后，AVL 树为" << std::endl;
  printTree(tree.root);
}

int main() {
  /* 初始化空 AVL 树 */
  AVLTree avlTree;

  /* 插入节点 */
  // 请关注插入节点后，AVL 树是如何保持平衡的
  testInsert(avlTree, 1);
  testInsert(avlTree, 2);
  testInsert(avlTree, 3);
  testInsert(avlTree, 4);
  testInsert(avlTree, 5);
  testInsert(avlTree, 8);
  testInsert(avlTree, 7);
  testInsert(avlTree, 9);
  testInsert(avlTree, 10);
  testInsert(avlTree, 6);

  /* 插入重复节点 */
  testInsert(avlTree, 7);

  /* 删除节点 */
  // 请关注删除节点后，AVL 树是如何保持平衡的
  testRemove(avlTree, 8); // 删除度为 0 的节点
  testRemove(avlTree, 5); // 删除度为 1 的节点
  testRemove(avlTree, 4); // 删除度为 2 的节点

  /* 查询节点 */
  TreeNode *node = avlTree.search(7);
  std::cout << "\n查找到的节点对象为 " << node << "，节点值 = " << node->val
            << std::endl;
}
